### ISSUE TITLE
Synchronization fails with IOException while parsing EXAMINE response

### ISSUE THREAD

**Issue by lenerd** [k9mail/k-9#4925]
### Expected behavior
K-9 Mail should synchronize all my mail accounts.

### Actual behavior
Since I installed v5.717 from F-Droid (used 5.600 before), I stopped receiving mails from one of my accounts. The other two work fine.

It looks like the parser dies when processing the response to `EXAMINE` (the full log is attached below):
```
08-28 16:08:52.542 11571 11595 V ImapConnection: conn88225688>>> 2 EXAMINE "INBOX"
08-28 16:08:52.560 11571 11595 E ImapFolder: IOException for AccountName:INBOX/MessagingController/conn88225688
08-28 16:08:52.560 11571 11595 E ImapFolder: java.io.IOException: parseBareString(): (0004 ^D)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.parseBareString(ImapResponseParser.java:325)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.parseToken(ImapResponseParser.java:256)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.parseList(ImapResponseParser.java:297)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.parseToken(ImapResponseParser.java:231)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.readToken(ImapResponseParser.java:219)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.readTokens(ImapResponseParser.java:140)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.readUntaggedResponse(ImapResponseParser.java:69)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.readResponse(ImapResponseParser.java:40)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.readResponse(ImapResponseParser.java:28)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapResponseParser.readStatusResponse(ImapResponseParser.java:86)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapConnection.executeSimpleCommand(ImapConnection.java:752)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapConnection.executeSimpleCommand(ImapConnection.java:738)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapFolder.executeSimpleCommand(ImapFolder.kt:95)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapFolder.internalOpen(ImapFolder.kt:131)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.mail.store.imap.ImapFolder.open(ImapFolder.kt:100)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.backend.imap.ImapSync.synchronizeMailboxSynchronous(ImapSync.java:96)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.backend.imap.ImapSync.sync(ImapSync.java:52)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.backend.imap.ImapBackend.sync(ImapBackend.java:114)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.controller.MessagingController.syncFolder(MessagingController.java:640)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.controller.MessagingController.synchronizeMailboxSynchronous(MessagingController.java:609)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.controller.MessagingController.lambda$synchronizeMailbox$3$MessagingController(MessagingController.java:596)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.controller.-$$Lambda$MessagingController$Xx9NETpzWcoXEDzz-o_rzXywOEM.run(lambda)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.controller.MessagingController.runInBackground(MessagingController.java:218)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.controller.MessagingController.access$000(MessagingController.java:112)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at com.fsck.k9.controller.MessagingController$1.run(MessagingController.java:161)
08-28 16:08:52.560 11571 11595 E ImapFolder:    at java.lang.Thread.run(Thread.java:761)
08-28 16:08:52.563 11571 11595 E ImapSync: synchronizeMailbox
```

When I manually open an IMAP connection to the server via ncat and run EXAMINE, it looks like this:
```
* OK [CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE XLIST AUTH=PLAIN] Dovecot ready.
a authenticate plain
+ 
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
a OK [CAPABILITY IMAP4rev1 LITERAL+ SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS THREAD=ORDEREDSUBJECT MULTIAPPEND URL-PARTIAL CATENATE UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS BINARY MOVE XLIST SPECIAL-USE QUOTA] Logged in
" EXAMINE INBOX
* FLAGS (\Answered \Flagged \Deleted \Seen \Draft NonJunk $label1 $Forwarded  Junk)
* OK [PERMANENTFLAGS ()] Read-only mailbox.
* 1785 EXISTS
* 0 RECENT
* OK [UIDVALIDITY 1507645949] UIDs valid
* OK [UIDNEXT 3429] Predicted next UID
" OK [READ-ONLY] Examine completed (0.000 + 0.000 secs)
```
Please let me know how I can help with debugging this.

### Steps to reproduce
I don't know how without having an account on the mail server.

### Environment
K-9 Mail version: v5.717 from F-Droid

Android version: 7.1.2

Account type (IMAP, POP3, WebDAV/Exchange): IMAP

Log: [k9-log-redacted.txt](https://github.com/k9mail/k-9/files/5142878/k9-log-redacted.txt)

**Comment by cketti** [comment]
The exception is thrown because the control character with the code point 0x0004 was found. I'm guessing it's the character right before "Junk" in the line
```
* FLAGS (\Answered \Flagged \Deleted \Seen \Draft NonJunk $label1 $Forwarded  Junk)
```

This is not a valid character for an "atom" token and it's a bug if the server sends such a response. 

I don't think Dovecot allows setting such a keyword via IMAP. So if I had to guess I'd say this is a random bit flip in the saved metadata (the code point for `$` is 0x44 (01000100) vs 0x04 (00000100)) and Dovecot doesn't re-check the contents of the keyword when reading it from disk.

Chances are this will go away if you delete the message with the affected keyword. It's probably still worth reporting this bug to Dovecot.

**Comment by lenerd** [comment]
Hi @cketti,

thanks a lot! I would not have noticed this and thought 0004 was some error code.

You were right, there is actually some junk before "Junk" ...
```
00000220: 466f 7277 6172 6465 6420 8080 8085 0420  Forwarded ..... 
00000230: 4a75 6e6b 290d 0a2a 204f 4b20 5b50 4552  Junk)..* OK [PER
```

I used Python's imaplib to go through all mails and retrieve their flags (of course, the server would not let me search for invalid flags directly), and have deleted the offending mail.

The `EXAMINE INBOX` command, however, still lists the bogus Junk flag, but not the benign one. Hence, K-9 still chokes. When fetching a test mail that I flagged as Junk in Thunderbird, I get the correct flag: ('OK', [b'718 (UID 3429 FLAGS (\\Seen Junk))'])

I guess I need to talk to the admins and maybe the Dovecot developers.
